lib/Sema/DerivedConformanceCodable.cpp

```diff
@@ -353,8 +358,21 @@ static EnumDecl *synthesizeCodingKeysEnum(DerivedConformance &derived) {
       }
 
       case DoesNotConform:
+        // for accurate diagnostic, we show TypeRepr if present for IUO
+        TypeLoc typeLoc = TypeLoc();
+        if(varDecl->isImplicitlyUnwrappedOptional()){
+            typeLoc = {
+                varDecl->getTypeReprOrParentPatternTypeRepr(),
+                varDecl->getType(),
+            };
+        } else {
+            typeLoc = {
+                NULL,
+                varDecl->getType(),
+            };
+        }
```

Comment:
Given we're doing this in three places in this function, I think it would be a good idea to extract and reuse:

```c++
static TypeLoc getTypeLocForDiagnostics(VarDecl *VD) {
  auto varTy = VD->getType();
  if (VD->isImplicitlyUnwrappedOptional()) {
    return TypeLoc(VD->getTypeReprOrParentPatternTypeRepr(), varTy);
  }
  return TypeLoc::withoutLoc(varTy);
}
```

Reply:
I don't think this needs to be a utility. Its ideal form is four lines

```
TypeLoc typeLoc = {
  varDecl->getTypeReprOrParentPatternTypeRepr(),
 varDecl->getType(),
};
```